torch/csrc/distributed/rpc/rref_context.h

```diff
@@ -138,6 +141,20 @@ class RRefContext {
   mutable std::mutex mutex_;
   // Keep OwnerRRefs alive until there is no living UserRRefs.
   std::unordered_map<RRefId, std::shared_ptr<RRef>, RRefId::Hash> owners_;
+  // Conditional variables to block getOwnerRRef() calls until the corresponding
+  // OwnerRRef has been created and inserted into the owners_ map. The method
+  // getOwnerRRef() is triggered by rref.to_here() messages. The reason for
+  // having this CV is because rref.to_here() message and rpc.remote message may
+  // arrive in any order, and to_here() can only be served when the RRef value
+  // is ready. In the previous version, we used to block the to_here() call by
+  // waiting on the CV member variable in OwnerRRef. However, that means the
+  // to_here() call has to first create the OwnerRRef, which would require
+  // knowing the IValue type when if we want to make RRef an IValue. Instead of
+  // sending serialized TypePtr in every to_here() message, we decided to only
+  // create OwnerRRef when processing remote calls.
+  // TODO: As OwnerRRef::getValue() is always called after
+  // OwnerRRef::setValue(), we should be able to remove the CV from OwnerRRef.
+  std::unordered_map<RRefId, std::condition_variable, RRefId::Hash> ownerCVs_;
```

Comment:
Just to ask: would it be a lesser evil to just use a single condvar, shared across everybody in getOwnerRRef waiting on an rref? Realistically, how many are going to be waiting at once in condvar::wait?

The cost would be some spurious wakeups, but upon waking up it's just a hash table loop, and back to sleep. The extra work ends having an N^2 factor with respect to num waiting, so nothing we want to do if there are thousands waiting. But it might be better than either leaking memory forever, or the complexity of managing memory lifetime here?

Alternatively, if that approach is too simple, another approach would be to keep a  std::vector<std::condition_variable> of a small fixed size (say 16 or whatever), and share among rrefs by mod sharding (e.g. rrefId % num_vector_condvars)

Reply:
Thanks for commenting!

> Realistically, how many are going to be waiting at once in condvar::wait?

Whenever there are OwnerRRefs used before creation would have one waiting for this CV. This probably won't be too many and should be a transient stage for most cases. I agree this might be an overkill to keep this CV live for the entire lifetime of an OwnerRRef. Let me make changes. Thanks!

Reply:
Great, looks somewhat simpler now!